examples/python/learning_pytorch.py

```diff
@@ -48,230 +47,250 @@
+            self.q_net = DuelQNet(action_size).to(DEVICE)
+            self.target_net = DuelQNet(action_size).to(DEVICE)
 
-    s1 = preprocess(game.get_state().screen_buffer)
+        self.opt = optim.SGD(self.q_net.parameters(), lr=self.lr)
 
-    # With probability eps make a random action.
-    eps = exploration_rate(epoch)
-    if random() <= eps:
-        a = randint(0, len(actions) - 1)
-    else:
-        # Choose the best action according to the network.
-        s1 = s1.reshape([1, 1, resolution[0], resolution[1]])
-        a = get_best_action(s1)
-    reward = game.make_action(actions[a], frame_repeat)
+    def get_action(self, state):
+        if np.random.uniform() < self.epsilon:
+            return random.choice(range(self.action_size))
+        else:
+            state = np.expand_dims(state, axis=0)
+            action = torch.argmax(self.q_net(state)).item()
+            return action
 
-    isterminal = game.is_episode_finished()
-    s2 = preprocess(game.get_state().screen_buffer) if not isterminal else None
+    def update_target_net(self):
+        self.target_net.load_state_dict(self.q_net.state_dict())
 
-    # Remember the transition that was just experienced.
-    memory.add_transition(s1, a, s2, isterminal, reward)
+    def append_memory(self, state, action, reward, next_state, done):
+        self.memory.append((state, action, reward, next_state, done))
 
-    learn_from_memory()
+    def train(self):
+        batch = random.sample(self.memory, self.batch_size)
+        batch = np.array(batch, dtype=object)
 
+        states = np.stack(batch[:, 0]).astype(float)
+        actions = batch[:, 1].astype(int)
+        rewards = batch[:, 2].astype(float)
+        next_states = np.stack(batch[:, 3]).astype(float)
+        dones = batch[:, 4].astype(bool)
+        not_dones = ~dones
 
-# Creates and initializes ViZDoom environment.
-def initialize_vizdoom(config_file_path):
-    print("Initializing doom...")
-    game = DoomGame()
-    game.load_config(config_file_path)
-    game.set_window_visible(False)
-    game.set_mode(Mode.PLAYER)
-    game.set_screen_format(ScreenFormat.GRAY8)
-    game.set_screen_resolution(ScreenResolution.RES_640X480)
-    game.init()
-    print("Doom initialized.")
-    return game
+        row_idx = np.arange(self.batch_size)  # used for indexing the batch
 
+        # value of the next states with double q learning
+        with torch.no_grad():
+            idx = row_idx, np.argmax(self.q_net(next_states).cpu().data.numpy(), 1)
+            next_state_values = self.target_net(next_states).cpu().data.numpy()[idx]
+            next_state_values = next_state_values[not_dones]
 
-if __name__ == '__main__':
-    # Create Doom instance
-    game = initialize_vizdoom(config_file_path)
+        # this defines y = r + discount * max_a q(s', a)
+        q_targets = rewards.copy()
+        q_targets[not_dones] += self.discount * next_state_values
+        q_targets = torch.from_numpy(q_targets).float()
```

Comment:
Missing `.to(DEVICE)` (does not run for me without it on CUDA).

Reply:
Done